text/src/tail.rs

```diff
@@ -0,0 +1,236 @@
+use clap::Parser;
+use gettextrs::{bind_textdomain_codeset, textdomain};
+use notify_debouncer_full::new_debouncer;
+use notify_debouncer_full::notify::event::{ModifyKind, RemoveKind};
+use notify_debouncer_full::notify::{EventKind, RecursiveMode, Watcher};
+use plib::PROJECT_NAME;
+use std::fs;
+use std::fs::File;
+use std::io::{self, BufRead, BufReader, Read, Seek, SeekFrom, Write};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::time::Duration;
+
+/// Wrapper type for isize that defaults to negative values if no sign is provided
+#[derive(Debug, Clone)]
+struct SignedIsize(isize);
+
+impl FromStr for SignedIsize {
+    type Err = <isize as FromStr>::Err;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.starts_with('-') || s.starts_with('+') {
+            Ok(SignedIsize(s.parse()?))
+        } else {
+            Ok(SignedIsize(-s.parse::<isize>()?))
+        }
+    }
+}
+
+/// tail - copy the last part of a file
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// The number of lines to print from the end of the file
+    #[arg(short = 'n')]
+    lines: Option<isize>,
+
+    /// The number of bytes to print from the end of the file
+    #[arg(short = 'c')]
+    bytes: Option<SignedIsize>,
+
+    /// Output appended data as the file grows
+    #[arg(short = 'f')]
+    follow: bool,
+
+    /// The file to read
+    file: Option<PathBuf>,
+}
+
+impl Args {
+    /// Validates the command-line arguments to ensure they meet the required constraints.
+    ///
+    /// # Returns
+    /// * `Ok(())` if arguments are valid.
+    /// * `Err(String)` if arguments are invalid, with an error message describing the issue.
+    fn validate_args(&mut self) -> Result<(), String> {
+        // Check if conflicting options are used together
+        if self.bytes.is_some() && self.lines.is_some() {
+            return Err("Options '-c' and '-n' cannot be used together".to_string());
+        }
+
+        if self.bytes.is_none() && self.lines.is_none() {
+            self.lines = Some(-10);
+        }
+
+        Ok(())
+    }
+}
+
+/// Prints the last `n` lines from the given buffered reader.
+///
+/// # Arguments
+/// * `reader` - A buffered reader to read lines from.
+/// * `n` - The number of lines to print from the end. Negative values indicate counting from the end.
+
+fn print_last_n_lines<R: BufRead>(reader: R, n: isize) -> Result<(), String> {
+    let lines: Vec<_> = reader.lines().map_while(Result::ok).collect();
+
```

Comment:
This appears to read the entire file into memory.  That is definitely unwanted behavior... the file might be several gigabytes in size!  

The tail utility should keep **at maximum** $N lines of content in memory.